Make gen_annotated_args support kwargs

guangy10 opened the pull request:
This PR is to address the issue seeing in PR #97417 where the newly added op requires `kwargs`, however, currently tools/autograd/gen_annotated_fn_args.py does not support `kwargs`, only `func_args` are generated for test_overrides.py.

The PR adds a new field "is_kwargs" to each argument indicating whether it's a `kwargs` or not. See example: 
```
annotated_args = {
    torch._C._VariableFunctions._cast_Byte: [{'is_kwarg_only': 'False', 'name': 'self', 'simple_type': 'Tensor'}],
    ...
```

The full comparison of the generated file `annotated_fn_args.py` can be found here:
  - **Before**: [P681991116](https://www.internalfb.com/phabricator/paste/view/P681991116)
  - **After**: [P681994218](https://www.internalfb.com/intern/paste/P681994218/)

Differential Revision: D44698310

guangy10 commented on tools/autograd/gen_annotated_fn_args.py:
```diff
@@ -77,17 +77,34 @@ def gen_annotated(
 def gen_annotated_args(f: NativeFunction) -> str:
+    def _get_kwargs_func_exclusion_list():
+        # functions that don't work with kwargs in test_overrides.py
+        return [
+            "diagonal",
+            "round_",
+            "round",
+            "scatter_",
+        ]
```
End of day I'm not able to get all ops works with `kwargs` hence creating this exclusion list to move forward. The goal of this PR is to unblock shipping #97417

albanD commented:
Hey!
Could you share example updated generated code please?

guangy10 commented:
> Could you share example updated generated code please?

@albanD Yes. I updated the PR summary to include the example and full comparison of the generated file.

ezyang commented on test/test_overrides.py:
```diff
@@ -629,53 +629,65 @@ def instance_gen():
+                    func = func.__get__(instance_gen())
+                    continue
+                arg_to_add = _simple_type_parser(func, arg)
+                if 'is_kwargs' in arg and arg['is_kwargs'] == str(True):
```
bikeshed: `is_kwarg_only`

ezyang replied:
How come is_kwargs isn't just a bool True?

albanD replied:
+1 on `is_kwarg_only`

guangy10 replied:
Done

ezyang commented on test/test_overrides.py:
```diff
@@ -629,53 +629,65 @@ def instance_gen():
                 # Guess valid input to aten function based on type of argument
                 t = arg['simple_type']
                 if t.endswith('?'):
                     t = t[:-1]
```
How come you need to do this part twice now?

guangy10 replied:
Addressed

ezyang commented on tools/autograd/gen_annotated_fn_args.py:
```diff
@@ -77,17 +77,36 @@ def gen_annotated(
+            if arg.default is not None:
+                continue
+            out_arg: Dict[str, Any] = {}
+            out_arg["is_kwargs"] = str(is_kwargs)
```
just assign this as `is_kwarg_only`?

ezyang reviewed (approved):
This seems fine, please check the bikesheds

albanD reviewed (commented):
Thanks for the updated version. The diff looks good.
No other comments beyond Ed's ones.

guangy10 committed:
Make gen_annotated_args support kwargs (#98396)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/98396

Test Plan:
`buck2 test mode/dev-nosan //caffe2/test:overrides`

https://www.internalfb.com/intern/testinfra/testrun/6755399588715296

Differential Revision: D44698310

fbshipit-source-id: f49e6bd260ef966215f8ea297b98be4e7ffce537

guangy10 commented:
@pytorchbot

guangy10 commented:
@pytorchbot label "topic: not user facing"

guangy10 commented:
Irrelevant CI failures. 

@pytorchbot merge -ic
